Doesn't work, java error

Boot, first translate- ok, all work, 
![Screenshot_2023-03-11-01-55-34-797_snoddasmannen galimulator](https://user-images.githubusercontent.com/69048938/224443663-1c08908a-9ab7-42b8-a989-ff803fc14626.jpg)
second translate- error.

2023-09-17 14:38:38.497 27850-28027 ScreenExtractor         tw.firemaples.onscreenocr_debug      W  [DefaultDispatcher-worker-3]
                                                                                                    java.lang.NullPointerException: Attempt to invoke interface method 'boolean java.util.Set.add(java.lang.Object)' on a null object reference
                                                                                                    	at android.os.Parcel.createExceptionOrNull(Parcel.java:2431)
                                                                                                    	at android.os.Parcel.createException(Parcel.java:2409)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:2392)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:2334)
                                                                                                    	at android.media.projection.IMediaProjection$Stub$Proxy.start(IMediaProjection.java:235)
                                                                                                    	at android.media.projection.MediaProjection.<init>(MediaProjection.java:59)
                                                                                                    	at android.media.projection.MediaProjectionManager.getMediaProjection(MediaProjectionManager.java:119)
                                                                                                    	at tw.firemaples.onscreenocr.screenshot.ScreenExtractor$doCaptureScreen$2.invokeSuspend(ScreenExtractor.kt:113)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: android.os.RemoteException: Remote stack trace:
                                                                                                    	at com.android.server.media.projection.MediaProjectionManagerService$MediaProjection.start(MediaProjectionManagerService.java:553)
                                                                                                    	at android.media.projection.IMediaProjection$Stub.onTransact(IMediaProjection.java:137)
                                                                                                    	at android.os.Binder.execTransactInternal(Binder.java:1187)
                                                                                                    	at android.os.Binder.execTransact(Binder.java:1146)
2023-09-17 14:38:38.499 27850-27850 FloatingStateManager    tw.firemaples.onscreenocr_debug      D  [main]
                                                                                                    java.lang.NullPointerException: Attempt to invoke interface method 'boolean java.util.Set.add(java.lang.Object)' on a null object reference
                                                                                                    	at android.os.Parcel.createExceptionOrNull(Parcel.java:2431)
                                                                                                    	at android.os.Parcel.createException(Parcel.java:2409)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:2392)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:2334)
                                                                                                    	at android.media.projection.IMediaProjection$Stub$Proxy.start(IMediaProjection.java:235)
                                                                                                    	at android.media.projection.MediaProjection.<init>(MediaProjection.java:59)
                                                                                                    	at android.media.projection.MediaProjectionManager.getMediaProjection(MediaProjectionManager.java:119)
                                                                                                    	at tw.firemaples.onscreenocr.screenshot.ScreenExtractor$doCaptureScreen$2.invokeSuspend(ScreenExtractor.kt:113)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: android.os.RemoteException: Remote stack trace:
                                                                                                    	at com.android.server.media.projection.MediaProjectionManagerService$MediaProjection.start(MediaProjectionManagerService.java:553)
                                                                                                    	at android.media.projection.IMediaProjection$Stub.onTransact(IMediaProjection.java:137)
                                                                                                    	at android.os.Binder.execTransactInternal(Binder.java:1187)
                                                                                                    	at android.os.Binder.execTransact(Binder.java:1146)



![image](https://github.com/firemaples/EverTranslator/assets/69048938/4df85fad-ad01-4252-8ef9-bd00ea4fc9b5)
![image](https://github.com/firemaples/EverTranslator/assets/69048938/baaa8c05-5cc4-48a7-ae94-07a2676cb135)

Add 5 internal lines:


[DefaultDispatcher-worker-3]
                                                                                                    java.lang.NullPointerException: Attempt to invoke interface method 'boolean java.util.Set.add(java.lang.Object)' on a null object reference
                                                                                                    	at android.os.Parcel.createExceptionOrNull(Parcel.java:2431)
                                                                                                    	at android.os.Parcel.createException(Parcel.java:2409)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:2392)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:2334)
                                                                                                    	at android.media.projection.IMediaProjection$Stub$Proxy.start(IMediaProjection.java:235)
                                                                                                    	at android.media.projection.MediaProjection.<init>(MediaProjection.java:59)
                                                                                                    	at android.media.projection.MediaProjectionManager.getMediaProjection(MediaProjectionManager.java:119)
                                                                                                    	at tw.firemaples.onscreenocr.screenshot.ScreenExtractor$doCaptureScreen$2.invokeSuspend(ScreenExtractor.kt:113)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    Caused by: android.os.RemoteException: Remote stack trace:
                                                                                                    	at com.android.server.media.projection.MediaProjectionManagerService$MediaProjection.start(MediaProjectionManagerService.java:553)
                                                                                                    	at android.media.projection.IMediaProjection$Stub.onTransact(IMediaProjection.java:137)
                                                                                                    	at android.os.Binder.execTransactInternal(Binder.java:1187)
                                                                                                    	at android.os.Binder.execTransact(Binder.java:1146)


![image](https://github.com/firemaples/EverTranslator/assets/69048938/ceb6ce56-9940-4f55-914f-29750f862806)

--- a/main/src/main/java/tw/firemaples/onscreenocr/pages/launch/LaunchActivity.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/pages/launch/LaunchActivity.kt
@@ -7,6 +7,7 @@ import androidx.appcompat.app.AppCompatActivity
 import tw.firemaples.onscreenocr.databinding.ActivityLaunchBinding
 import tw.firemaples.onscreenocr.remoteconfig.RemoteConfigManager
 import tw.firemaples.onscreenocr.utils.AdManager
+import tw.firemaples.onscreenocr.utils.DeviceInfoChecker
 
 class LaunchActivity : AppCompatActivity() {
 
@@ -29,5 +30,7 @@ class LaunchActivity : AppCompatActivity() {
 //        MoPubAdManager.loadPermissionPageBanner(this, findViewById(R.id.ad_permissionPage))
 
         RemoteConfigManager.tryFetchNew()
+
+        DeviceInfoChecker.check()
     }
 }

--- a/main/src/main/java/tw/firemaples/onscreenocr/pages/launch/permissions/PermissionCaptureScreenFragment.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/pages/launch/permissions/PermissionCaptureScreenFragment.kt
@@ -11,6 +11,7 @@ import androidx.activity.result.contract.ActivityResultContracts
 import androidx.fragment.app.Fragment
 import tw.firemaples.onscreenocr.databinding.PermissionCaptureScreenFragmentBinding
 import tw.firemaples.onscreenocr.floatings.ViewHolderService
+import tw.firemaples.onscreenocr.pages.setting.SettingManager
 import tw.firemaples.onscreenocr.screenshot.ScreenExtractor
 import tw.firemaples.onscreenocr.utils.clickOnce
 
@@ -54,7 +55,10 @@ class PermissionCaptureScreenFragment : Fragment() {
         registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
             val intent = it.data
             if (it.resultCode == Activity.RESULT_OK && intent != null) {
-                ScreenExtractor.onMediaProjectionGranted(intent, true)
+                ScreenExtractor.onMediaProjectionGranted(
+                    intent = intent,
+                    keepMediaProjection = SettingManager.keepMediaProjectionResources,
+                )
                 startService()
             }
         }

--- a/main/src/main/java/tw/firemaples/onscreenocr/pages/launch/permissions/PermissionCaptureScreenFragment.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/pages/launch/permissions/PermissionCaptureScreenFragment.kt
@@ -54,7 +54,7 @@ class PermissionCaptureScreenFragment : Fragment() {
         registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
             val intent = it.data
             if (it.resultCode == Activity.RESULT_OK && intent != null) {
-                ScreenExtractor.onMediaProjectionGranted(intent)
+                ScreenExtractor.onMediaProjectionGranted(intent, true)
                 startService()
             }
         }

--- a/main/src/main/java/tw/firemaples/onscreenocr/pages/setting/SettingFragment.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/pages/setting/SettingFragment.kt
@@ -8,7 +8,9 @@ import android.text.InputType
 import androidx.preference.EditTextPreference
 import androidx.preference.Preference
 import androidx.preference.PreferenceFragmentCompat
+import androidx.preference.SwitchPreference
 import tw.firemaples.onscreenocr.R
+import tw.firemaples.onscreenocr.utils.Toaster
 import tw.firemaples.onscreenocr.utils.Utils
 
 class SettingFragment : PreferenceFragmentCompat() {
@@ -17,7 +19,10 @@ class SettingFragment : PreferenceFragmentCompat() {
         get() = findPreference("pref_bypass_battery_optimization")
 
     private val myMemoryEmail: EditTextPreference?
-        get() = findPreference("pref_mymemory_email")
+        get() = findPreference(SettingManager.PREF_MYMEMORY_EMAIL)
+
+    private val keepMediaProjectionResources: SwitchPreference?
+        get() = findPreference(SettingManager.PREF_KEEP_MEDIA_PROJECTION_RESOURCES)
 
     override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
         setPreferencesFromResource(R.xml.perference, rootKey)
@@ -31,6 +36,11 @@ class SettingFragment : PreferenceFragmentCompat() {
 
         myMemoryEmail?.setOnBindEditTextListener {
             it.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS
+        }
+
+        keepMediaProjectionResources?.setOnPreferenceChangeListener { _, _ ->
+            Toaster.show(getString(R.string.msg_please_restart_the_app_to_take_effect))
+            true
         }
     }
 

--- a/main/src/main/java/tw/firemaples/onscreenocr/pages/setting/SettingManager.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/pages/setting/SettingManager.kt
@@ -1,6 +1,7 @@
 package tw.firemaples.onscreenocr.pages.setting
 
 import android.content.Context
+import androidx.core.content.edit
 import androidx.core.util.PatternsCompat
 import androidx.preference.PreferenceManager
 import kotlinx.coroutines.CoroutineScope
@@ -60,8 +61,11 @@ object SettingManager {
     val opaquePercentageToFadeOut: Float
         get() = preferences.getInt(PREF_OPAQUE_PERCENTAGE, 60).toFloat() / 100f
 
-    val keepMediaProjectionResources: Boolean
+    var keepMediaProjectionResources: Boolean
         get() = preferences.getBoolean(PREF_KEEP_MEDIA_PROJECTION_RESOURCES, false)
+        set(value) {
+            preferences.edit { putBoolean(PREF_KEEP_MEDIA_PROJECTION_RESOURCES, value) }
+        }
 
     val enableUnrecommendedLangItems: Boolean
         get() = preferences.getBoolean(PREF_ENABLE_UNRECOMMENDED_LANG_ITEMS, false)

--- a/main/src/main/java/tw/firemaples/onscreenocr/pages/setting/SettingManager.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/pages/setting/SettingManager.kt
@@ -20,6 +20,8 @@ object SettingManager {
     private const val PREF_FADE_OUT_AFTER_SECONDS = "pref_fade_out_after_seconds"
     private const val PREF_OPAQUE_PERCENTAGE = "pref_opaque_percentage"
 
+    const val PREF_KEEP_MEDIA_PROJECTION_RESOURCES = "pref_keep_media_projection_resources"
+
     private const val PREF_ENABLE_UNRECOMMENDED_LANG_ITEMS = "pref_enable_unrecommended_lang_items"
     private const val PREF_TIMEOUT_FOR_CAPTURING_SCREEN = "pref_timeout_for_capturing_screen"
     private const val PREF_TEXT_BLOCK_JOINER = "pref_text_block_joiner"
@@ -34,7 +36,7 @@ object SettingManager {
     private const val PREF_SAVE_LAST_SELECTION_AREA = "pref_save_last_selection_area"
     private const val PREF_EXIT_APP_WHILE_SPEN_INSERTED = "pref_exit_app_while_spen_inserted"
 
-    private const val PREF_MYMEMORY_EMAIL = "pref_mymemory_email"
+    const val PREF_MYMEMORY_EMAIL = "pref_mymemory_email"
 
     private val DEFAULT_JOINER = TextBlockJoiner.Space
 
@@ -57,6 +59,9 @@ object SettingManager {
 
     val opaquePercentageToFadeOut: Float
         get() = preferences.getInt(PREF_OPAQUE_PERCENTAGE, 60).toFloat() / 100f
+
+    val keepMediaProjectionResources: Boolean
+        get() = preferences.getBoolean(PREF_KEEP_MEDIA_PROJECTION_RESOURCES, false)
 
     val enableUnrecommendedLangItems: Boolean
         get() = preferences.getBoolean(PREF_ENABLE_UNRECOMMENDED_LANG_ITEMS, false)

--- a/main/src/main/java/tw/firemaples/onscreenocr/pref/AppPref.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/pref/AppPref.kt
@@ -52,4 +52,6 @@ object AppPref : KotprefModel() {
     val favoriteTranslationLang: MutableSet<String> by stringSetPref(default = mutableSetOf("en"))
 
     var displaySelectedTextOnResultWindow: Boolean by booleanPref(default = false)
+
+    var lastDeviceInfo: String by stringPref()
 }

--- a/main/src/main/java/tw/firemaples/onscreenocr/screenshot/ScreenExtractor.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/screenshot/ScreenExtractor.kt
@@ -39,6 +39,8 @@ object ScreenExtractor {
     private val context: Context by lazy { Utils.context }
     private val logger: Logger by lazy { Logger(ScreenExtractor::class) }
 
+    private var keepMediaProjection: Boolean = true
+
     private var mediaProjectionIntent: Intent? = null
 
     val isGranted: Boolean
@@ -50,15 +52,18 @@ object ScreenExtractor {
         Handler(thread.looper)
     }
 
+    private var projection: MediaProjection? = null
+    private var imageReader: ImageReader? = null
     private var virtualDisplay: VirtualDisplay? = null
-//    private val screenshotDir: File by lazy { File(context.cacheDir, Constants.PATH_SCREENSHOT) }
-//    private val screenshotFile: File by lazy { File(screenshotDir, "screenshot.jpg") }
+    private var lastScreenSize: Point = Point()
 
     private val screenDensityDpi: Int
         get() = UIUtils.displayMetrics.densityDpi
 
-    fun onMediaProjectionGranted(intent: Intent) {
+    fun onMediaProjectionGranted(intent: Intent, keepMediaProjection: Boolean) {
+        releaseAllResources()
         mediaProjectionIntent = intent.clone() as Intent
+        this.keepMediaProjection = keepMediaProjection
     }
 
     fun release() {
@@ -84,10 +89,6 @@ object ScreenExtractor {
         }
     }
 
-    private var projection: MediaProjection? = null
-    private var imageReader: ImageReader? = null
-    private var image: Image? = null
-
     @SuppressLint("WrongConstant")
     @Throws(
         IllegalStateException::class,
@@ -97,11 +98,13 @@ object ScreenExtractor {
     private suspend fun doCaptureScreen(): Bitmap {
         var bitmap: Bitmap
         withContext(Dispatchers.Default) {
-            releaseAllResources()
+            if (!keepMediaProjection) {
+                releaseAllResources()
+            }
 
             val mpIntent = mediaProjectionIntent
             if (mpIntent == null) {
-                logger.warn("The mediaProjectionIntent is null: $mpIntent")
+                logger.warn("The mediaProjectionIntent is null")
                 throw IllegalStateException("The media projection intent is not initialized")
             }
 
@@ -109,56 +112,66 @@ object ScreenExtractor {
                 context.getSystemService(Context.MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
 
             try {
-                projection =
-                    mpManager.getMediaProjection(Activity.RESULT_OK, mpIntent.clone() as Intent)
+                if (projection == null) {
+                    logger.debug("Create MediaProjection")
+                    projection =
+                        mpManager.getMediaProjection(Activity.RESULT_OK, mpIntent.clone() as Intent)
+                }
 
                 val projection = projection
                 if (projection == null) {
-                    logger.warn("Retrieve projection failed, projection: $projection")
+                    logger.warn("Retrieve projection failed, projection is null")
                     throw IllegalStateException("Retrieving media projection failed")
                 }
 
                 val screenSize = UIUtils.realSize
+                val screenConfigurationChanged = lastScreenSize != screenSize
+                lastScreenSize = screenSize
                 val width = screenSize.x
                 val height = screenSize.y
 
-                imageReader =
-                    ImageReader.newInstance(width, height, AppPref.imageReaderFormat, 2)
-                val imageReader = imageReader
-
-                if (imageReader == null) {
-                    logger.debug("The imageReader is null after initialized")
-                    releaseAllResources()
-                    throw IllegalStateException("No image reader initialized failed")
+                var imageReader: ImageReader? = this@ScreenExtractor.imageReader
+                if (imageReader == null || screenConfigurationChanged) {
+                    logger.debug("Create ImageReader")
+                    imageReader =
+                        ImageReader.newInstance(width, height, AppPref.imageReaderFormat, 2)
+                    imageReaderReady = false
+                    this@ScreenExtractor.imageReader = imageReader
                 }
 
-                virtualDisplay = projection.createVirtualDisplay(
-                    "screen-mirror",
-                    width, height, screenDensityDpi,
-                    DisplayManager.VIRTUAL_DISPLAY_FLAG_OWN_CONTENT_ONLY or DisplayManager.VIRTUAL_DISPLAY_FLAG_PUBLIC,
-                    imageReader.surface, null, null
-                )
+                var virtualDisplay: VirtualDisplay? = virtualDisplay
+                if (virtualDisplay == null) {
+                    logger.debug("Create VirtualDisplay")
+                    virtualDisplay = projection.createVirtualDisplay(
+                        "screen-mirror",
+                        width, height, screenDensityDpi,
+                        DisplayManager.VIRTUAL_DISPLAY_FLAG_OWN_CONTENT_ONLY or DisplayManager.VIRTUAL_DISPLAY_FLAG_PUBLIC,
+                        imageReader.surface, null, null
+                    )
+                    this@ScreenExtractor.virtualDisplay = virtualDisplay
+                } else if (screenConfigurationChanged) {
+                    // Update configuration changes
+                    logger.debug("screenConfigurationChanged: $lastScreenSize")
+                    virtualDisplay.resize(width, height, screenDensityDpi)
+                    virtualDisplay.surface = imageReader.surface
+                }
                 val captured = withTimeout(SettingManager.timeoutForCapturingScreen) {
                     logger.debug("awaitForBitmap")
                     imageReader.awaitForBitmap(screenSize)
                 }
 
-                virtualDisplay?.release()
-
                 if (captured == null) {
                     logger.warn("The captured image is null")
-                    releaseAllResources()
+                    if (!keepMediaProjection) {
+                        releaseAllResources()
+                    }
                     throw IllegalStateException("No image data found")
                 }
                 bitmap = captured
 
-                releaseAllResources()
-
                 logger.debug("Bitmap size: ${bitmap.width}x${bitmap.height}, screen size: ${width}x$height")
             } catch (e: Throwable) {
                 logger.warn(t = e)
-
-                releaseAllResources()
 
                 val message = e.message ?: e.localizedMessage
                 if (message != null) {
@@ -176,6 +189,10 @@ object ScreenExtractor {
                 }
 
                 throw e
+            } finally {
+                if (!keepMediaProjection) {
+                    releaseAllResources()
+                }
             }
         }
 
@@ -183,8 +200,9 @@ object ScreenExtractor {
     }
 
     private fun releaseAllResources() {
+        logger.debug("releaseAllResources()")
         try {
-            imageReader?.setOnImageAvailableListener(null, handler)
+            imageReader?.setOnImageAvailableListener(null, null)
         } catch (e: Exception) {
             // ignore exceptions
         }
@@ -193,16 +211,19 @@ object ScreenExtractor {
         } catch (e: Exception) {
             // ignore exceptions
         }
+        imageReader = null
+        try {
+            virtualDisplay?.release()
+        } catch (e: Exception) {
+            // ignore exceptions
+        }
+        virtualDisplay = null
         try {
             projection?.stop()
         } catch (e: Exception) {
             // ignore exceptions
         }
-        try {
-            image?.close()
-        } catch (e: Exception) {
-            // ignore exceptions
-        }
+        projection = null
     }
 
     @Throws(IllegalArgumentException::class)
@@ -240,8 +261,19 @@ object ScreenExtractor {
         return cropped
     }
 
+    private var imageReaderReady = false
     private suspend fun ImageReader.awaitForBitmap(screenSize: Point): Bitmap? =
         suspendCancellableCoroutine {
+            if (imageReaderReady) {
+                logger.debug("reader is ready, acquireLatestImage directly")
+                val image = acquireLatestImage()
+                val bitmap = image.decodeBitmap(screenSize)
+                image.close()
+                logger.debug("Latest bitmap: $bitmap")
+                it.resume(bitmap)
+                return@suspendCancellableCoroutine
+            }
+
             logger.debug("suspendCancellableCoroutine: Started")
             var counter = 0
             val resumed = AtomicBoolean(false)
@@ -259,6 +291,9 @@ object ScreenExtractor {
                         reader.setOnImageAvailableListener(null, null)
                         if (resumed.getAndSet(true))
                             return@setOnImageAvailableListener
+                        if (keepMediaProjection) {
+                            imageReaderReady = true
+                        }
                         it.resume(bitmap)
                     } else {
                         logger.info("Image is whole black, increase counter: $counter")

--- a/main/src/main/java/tw/firemaples/onscreenocr/utils/DeviceInfoChecker.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/utils/DeviceInfoChecker.kt
@@ -1,0 +1,42 @@
+package tw.firemaples.onscreenocr.utils
+
+import android.os.Build
+import tw.firemaples.onscreenocr.pages.setting.SettingManager
+import tw.firemaples.onscreenocr.pref.AppPref
+
+object DeviceInfoChecker {
+    private val logger: Logger by lazy { Logger(this::class) }
+    fun check() {
+        val manufacturer = Build.MANUFACTURER
+        val model = Build.MODEL
+        val os = Build.VERSION.SDK_INT
+
+        val deviceInfo = "$manufacturer,$model,$os"
+        logger.debug("DeviceInfo: $deviceInfo")
+
+        if (deviceInfo != AppPref.lastDeviceInfo) {
+            when {
+                manufacturer.contains("Xiaomi", ignoreCase = true) &&
+                        os >= Build.VERSION_CODES.S -> {
+                    keepMediaProjectionResources()
+                }
+
+                manufacturer.contains("Blackshark", ignoreCase = true) &&
+                        os >= Build.VERSION_CODES.S -> {
+                    keepMediaProjectionResources()
+                }
+
+                os >= 34 -> {
+                    keepMediaProjectionResources()
+                }
+            }
+        }
+
+        AppPref.lastDeviceInfo = deviceInfo
+    }
+
+    private fun keepMediaProjectionResources() {
+        logger.debug("keepMediaProjectionResources()")
+        SettingManager.keepMediaProjectionResources = true
+    }
+}

--- a/main/src/main/java/tw/firemaples/onscreenocr/utils/Toaster.kt
+++ b/main/src/main/java/tw/firemaples/onscreenocr/utils/Toaster.kt
@@ -3,15 +3,19 @@ package tw.firemaples.onscreenocr.utils
 import android.content.Context
 import android.widget.Toast
 import androidx.annotation.StringRes
+import java.lang.ref.WeakReference
 
 object Toaster {
     private val context: Context by lazy { Utils.context }
+    private var last: WeakReference<Toast>? = null
 
     fun show(msg: String) {
-        Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
+        last?.get()?.cancel()
+        last = WeakReference(Toast.makeText(context, msg, Toast.LENGTH_LONG).apply { show() })
     }
 
     fun show(@StringRes stringRes: Int) {
-        Toast.makeText(context, stringRes, Toast.LENGTH_SHORT).show()
+        last?.get()?.cancel()
+        last = WeakReference(Toast.makeText(context, stringRes, Toast.LENGTH_LONG).apply { show() })
     }
 }

--- a/main/src/main/res/values-zh-rCN/strings.xml
+++ b/main/src/main/res/values-zh-rCN/strings.xml
@@ -107,4 +107,9 @@
     <string name="pref_summary_mymemory_email">提供 MyMemory 翻译你的电子邮件以获取每日五万个字元的翻译额度。随处翻译将不会使用此电子邮件作为其他用途。</string>
     <string name="pref_title_mymemory_email">MyMemory 电子邮件</string>
     <string name="text_open_in_browser">在浏览器中开启</string>
+    <string name="msg_please_restart_the_app_to_take_effect">重启程式以生效</string>
+    <string name="pref_title_keep_media_projection_resources">维持萤幕截图相关资源</string>
+    <string name="pref_summary_on_keep_media_projection_resources">维持相关资源直到重启程式</string>
+    <string name="pref_summary_off_keep_media_projection_resources">每次使用后释放相关资源</string>
+    <string name="pref_screenshot">萤幕截图</string>
 </resources>

--- a/main/src/main/res/values-zh-rTW/strings.xml
+++ b/main/src/main/res/values-zh-rTW/strings.xml
@@ -107,4 +107,9 @@
     <string name="pref_summary_mymemory_email">提供 MyMemory 翻譯你的電子郵件以獲取每日五萬個字元的翻譯額度。隨處翻譯將不會使用此電子郵件作為其他用途。</string>
     <string name="pref_title_mymemory_email">MyMemory 電子郵件</string>
     <string name="text_open_in_browser">在瀏覽器中開啟</string>
+    <string name="msg_please_restart_the_app_to_take_effect">重啟程式以生效</string>
+    <string name="pref_title_keep_media_projection_resources">維持螢幕截圖相關資源</string>
+    <string name="pref_summary_on_keep_media_projection_resources">維持相關資源直到重啟程式</string>
+    <string name="pref_summary_off_keep_media_projection_resources">每次使用後釋放相關資源</string>
+    <string name="pref_screenshot">螢幕截圖</string>
 </resources>

--- a/main/src/main/res/values/strings.xml
+++ b/main/src/main/res/values/strings.xml
@@ -47,6 +47,10 @@
     <string name="pref_category_translation">Translation</string>
     <string name="pref_summary_mymemory_email">You can Provide a valid email to MyMemory translation for 50000 chars a day quota.\nEverTranslator will never use the email for other usages.</string>
     <string name="pref_title_mymemory_email">MyMemory email</string>
+    <string name="pref_screenshot">Screenshot</string>
+    <string name="pref_title_keep_media_projection_resources">Keep screenshot resources</string>
+    <string name="pref_summary_on_keep_media_projection_resources">Keep the resources until restart</string>
+    <string name="pref_summary_off_keep_media_projection_resources">Release the resources after used</string>
 
     <string name="menu_setting">Setting</string>
     <string name="menu_privacy_policy">Privacy Policy</string>
@@ -81,6 +85,7 @@
     <string name="msg_create_selection_area">Drag here to create a translation area</string>
     <string name="msg_image_reader_format_unmatched">Found temporary error, please try it again. (Image reader format)</string>
     <string name="msg_translator_provider_does_not_support_the_ocr_lang">This translation provider does not support the selected recognition language.</string>
+    <string name="msg_please_restart_the_app_to_take_effect">Please restart the app to take effect</string>
 
     <string name="title_version_history">Version History</string>
     <string name="title_error">Error</string>

--- a/main/src/main/res/xml/perference.xml
+++ b/main/src/main/res/xml/perference.xml
@@ -34,6 +34,15 @@
             app:showSeekBarValue="true" />
     </PreferenceCategory>
 
+    <PreferenceCategory android:title="@string/pref_screenshot">
+        <SwitchPreference
+            android:defaultValue="false"
+            android:key="pref_keep_media_projection_resources"
+            android:summaryOff="@string/pref_summary_off_keep_media_projection_resources"
+            android:summaryOn="@string/pref_summary_on_keep_media_projection_resources"
+            android:title="@string/pref_title_keep_media_projection_resources" />
+    </PreferenceCategory>
+
     <PreferenceCategory android:title="@string/pref_category_text_recognition">
         <SwitchPreference
             android:defaultValue="false"

